Source/ProgressionSystemRuntime/Private/Data/PSSaveGameData.cpp

```diff
@@ -0,0 +1,174 @@
+	}
+}
+
+// Unlocks the level specified by RowName if it exists in the saved progression rows.
+void UPSSaveGameData::UnlockLevelByName(FName RowName)
+{
+	// Using the operator [] on a TMap like SavedProgressionRowsInternal[RowName] can inadvertently create a new entry in the map if RowName does not exist
+	// Check if the row exists to avoid inadvertently creating a new entry
+	if (SavedProgressionRowsInternal.Contains(RowName))
+	{
+		FPSRowData& CurrentRowRef = SavedProgressionRowsInternal[RowName];
+		CurrentRowRef.IsLevelLocked = false;
+	}
+	else
+	{
+		// log an error for the case where RowName does not exist
+		UE_LOG(LogTemp, Warning, TEXT("UnlockLevelByName: Row name '%s' not found in the progression rows."), *RowName.ToString());
+	}
+}
+
+// Updates the current level's progression based on the end game state and proceeds to the next level if unlocked.
+void UPSSaveGameData::SavePoints(EEndGameState EndGameState)
+{
+	// Check if the current row exists in the map before attempting to update it
+	if (SavedProgressionRowsInternal.Contains(CurrentRowNameInternal))
+	{
+		FPSRowData& CurrentRowRef = SavedProgressionRowsInternal[CurrentRowNameInternal];
+		// Increase the current level's progression by the reward from the end game state
+		CurrentRowRef.CurrentLevelProgression += GetProgressionReward(EndGameState);
+
+		// Check if the current level progression has reached or surpassed the points needed to unlock
+		if (CurrentRowRef.CurrentLevelProgression >= CurrentRowRef.PointsToUnlock)
+		{
+			NextLevelProgressionRowData();  // Advance to the next level's progression data
+		}
+		UPSWorldSubsystem::Get().SaveDataAsync();  // Asynchronously save the updated data
+	}
+}
+
+// Advances to the next level progression row and unlocks it, if available, after the current row.
+void UPSSaveGameData::NextLevelProgressionRowData()
+{
+	bool bNextRowFound = false;
+
+	for (const TTuple<FName, FPSRowData>& KeyValue : SavedProgressionRowsInternal)
+	{
+		if (bNextRowFound)
+		{
+			UnlockLevelByName(KeyValue.Key);
+			break;
+		}
+
+		if (KeyValue.Key == CurrentRowNameInternal)
+		{
+			bNextRowFound = true;  // Indicate that the current row has been found
+		}
+	}
+}
+
+// Unlocks all levels and set maximum allowed progression points
+void UPSSaveGameData::UnlockAllLevels()
+{
+	for (TTuple<FName, FPSRowData>& KeyValue : SavedProgressionRowsInternal)
+	{
+		UnlockLevelByName(KeyValue.Key);
+		KeyValue.Value.CurrentLevelProgression = KeyValue.Value.PointsToUnlock;
+	}
+}
+
+// @h4rdmol - make function const
+// Retrieves the progression reward based on the end game state for the current level.
+float UPSSaveGameData::GetProgressionReward(EEndGameState EndGameState)
+{
+	// Verify that the current row exists in the map to prevent creating a new entry
+	if (!SavedProgressionRowsInternal.Contains(CurrentRowNameInternal))
+	{
+		UE_LOG(LogTemp, Warning, TEXT("GetProgressionReward: CurrentRowName '%s' not found."), *CurrentRowNameInternal.ToString());
```

Comment:
Getters with logs\ensures are not common. It's responsibility of callers to handle wrong result.
1. Please remove this log from your GetProgressionReward
2. Go to the source functions that calls `GetProgressionReward()` and in that original place do smth like next:
```
const float ProgressionReward = SaveData->GetProgressionReward();
if (ensureMsgf(ProgressionReward != 0.f, TEXT("ASSERT: [%i] %hs:\n'ProgressionReward can be null!"), __LINE__, __FUNCTION__))
{
	//... 
}
```

Reply:
agree, resolved: https://github.com/h4rdmol/ProgressionSystem/pull/1/commits/dbcfb11f4564eaad33c0a22d8ddb1b3293bb42b9
Resolved with adding checkf since row above checking the results and work with it